In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import statsmodels.api as sm
import seaborn as sns

In [19]:
df = pd.read_csv("CH14PR14.txt", sep='\s+', header=None, names=[ 'Y', 'X1', 'X2', 'X3'])
df.head()

Y  X1  X2  X3
0  0  59  52   0
1  0  61  55   1
2  1  82  51   0
3  0  51  70   0
4  0  53  70   0

In [20]:
X1= df['X1']
X2= df['X2']
X3= df['X3']
Y=df['Y']

### a. Find the maximum likelihood estimates of β0, β1, β2, and β3. State the fitted responsefunction.

In [21]:
import statsmodels.api as sm

X = sm.add_constant(df[['X1', 'X2', 'X3']])
print(X.head())
Y = df['Y']

logit_model = sm.GLM(Y, X, family=sm.families.Binomial())
logit_results = logit_model.fit()
print(logit_results.summary())

   const  X1  X2  X3
0    1.0  59  52   0
1    1.0  61  55   1
2    1.0  82  51   0
3    1.0  51  70   0
4    1.0  53  70   0
                 Generalized Linear Model Regression Results                  
Dep. Variable:                      Y   No. Observations:                  159
Model:                            GLM   Df Residuals:                      155
Model Family:                Binomial   Df Model:                            3
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -52.547
Date:                Sat, 10 Dec 2022   Deviance:                       105.09
Time:                        16:23:13   Pearson chi2:                     180.
No. Iterations:                     6   Pseudo R-squ. (CS):             0.1712
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
-----

In [22]:
logit_results.params

const   -1.177159
X1       0.072788
X2      -0.098986
X3       0.433975
dtype: float64

### Yi= -1.177159  +0.072788X1 + -0.098986X2 + 0.433975X3


### b. Obtain exp(b1), exp(b2), and exp(b3). Interpret these numbers.

In [23]:
b1=  0.072788
b2=  -0.098986
b3= 0.433975
betas=[b1,b2,b3]
ORs=[]

In [24]:
for b in betas: 
    OR= np.exp(b)
    ORs.append(OR)
print(ORs)

[1.075502506212986, 0.905755388510224, 1.5433802831270984]


### Conclusion: 
- The odds of receiving flu shot will increase by 1.35 times if the age of patien increases (X1) increases, as the exp(b1) = 1.075502506212986> 1
- The odds of recieving flu shot  will be less likely to happen by 0.905 times if the health awaréness (X2) increases, as the exp(b2)=  0.90575<1
- The odds of recieving flu shot will increase by 1.54 times if X3 increases, as the exp(b3)= 1.543>1

### c. What is the estimated probability that male clients aged 55 with a health awarenessindex of 60 will receive a flu shot?

In [27]:
# a= -1.177159 +0.072788X1 + -0.098986X2 + 0.433975X3
# => X3=1, aged 55=>X1=55, X2= 60
#calculate the probability corresponding to X
def logistic(a):
  return (np.exp(a) / (1 + np.exp(a)))

a= -1.177159 +0.072788*55 + -0.098986*60 + 0.433975*1 
pi = logistic(a)
print(pi)

0.06422370905902704


### The estimated probability that male clients age 55 with a health awareness index of 60 will recieve a flu shot= 0.06422370905902704

### d. Use the Wald test to determine whether X3, client gender, can be dropped from theregression model; use α = .05. State the alternatives, decision rule, and conclusion. Whatis the approximate P-value of the test?

In [31]:
logit_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:                      Y   No. Observations:                  159
Model:                            GLM   Df Residuals:                      155
Model Family:                Binomial   Df Model:                            3
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -52.547
Date:                Sat, 10 Dec 2022   Deviance:                       105.09
Time:                        16:41:28   Pearson chi2:                     180.
No. Iterations:                     6   Pseudo R-squ. (CS):             0.1712
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.1772      2.982     -0.395      0.693      -7.023       4.668
X1             0.0728      0.030      2.396      0.017       0.013       0.132
X2            -0.0990      0.033     -2.957      0.003      -0.165      -0.033
X3             0.4340      0.522      0.832      0.406      -0.589       1.457
==============================================================================
"""

In [ ]:
Z3=0.832

In [35]:
import statsmodels.formula.api as smf
import scipy.stats as stats
z_val= stats.norm.ppf(1-(0.05/2))
print(z_val)

1.959963984540054


### Conclusion: As Z3=0.832< Z_val=1.95, we can conclude null hypothesis H0, and claim that X3 can be excluded from the dataset. Besides that, p-value= 0.406 < alpha=0.05 => X3 can be dropped from the dataset